### ISSUE TITLE
canvas image is serialized in to_image before being updated

### ISSUE THREAD

**Issue by rocurley** [ggez/ggez#813]
**Describe the bug**
Turning an image into a vector and back again appears to clear the image.

**To Reproduce**
Here's a slightly modified version of the ggez basic project template, demonstrating the issue.
```rust
use ggez::event::{self, EventHandler};
use ggez::{graphics, Context, ContextBuilder, GameResult};
use graphics::Drawable;

fn main() {
    // Make a Context.
    let (mut ctx, mut event_loop) = ContextBuilder::new("my_game", "Cool Game Author")
        .build()
        .expect("aieee, could not create ggez context!");

    // Create an instance of your event handler.
    // Usually, you should provide it with the Context object to
    // use when setting your game up.
    let mut my_game = MyGame::new(&mut ctx);

    // Run!
    match event::run(&mut ctx, &mut event_loop, &mut my_game) {
        Ok(_) => println!("Exited cleanly."),
        Err(e) => println!("Error occured: {}", e),
    }
}

struct MyGame {
    // Your state here...
}

impl MyGame {
    pub fn new(_ctx: &mut Context) -> MyGame {
        // Load/create resources such as images here.
        MyGame {
		    // ...
		}
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        // Update code here...
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, graphics::WHITE);
        // Draw code here...
        let canvas = graphics::Canvas::with_window_size(ctx)?;
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_rectangle(
            ctx,
            graphics::DrawMode::fill(),
            graphics::Rect::new(10., 10., 100., 100.),
            graphics::BLACK,
        )?
        .draw(ctx, graphics::DrawParam::new())?;
        graphics::set_canvas(ctx, None);
        let image = canvas.into_inner();
        let image = break_everything(ctx, image)?; //Comment this line out to get expected behavior
        image.draw(ctx, graphics::DrawParam::new())?;
        graphics::present(ctx)
    }
}

fn break_everything(ctx: &mut Context, img: graphics::Image) -> GameResult<graphics::Image> {
    let w = img.width();
    let h = img.height();
    let pixels = img.to_rgba8(ctx)?;
    graphics::Image::from_rgba8(ctx, w, h, &pixels)
}
```

**Expected behavior**
A black rectangle to appear on the screen, instead of nothing. The rectangle would be flipped, as per #693, but would exist.

**Hardware and Software:**
 - ggez version: 0.51
 - OS: Linux (Mint, 19)
 - Graphics card: Intel HD Graphics 5500
 - Graphics card drivers: i915

**Comment by rocurley** [comment]
I ran into this bug when trying to un-flip the canvas image. As part of writing this repro, I realized that I can directly draw the canvas and not have it flipped. Given that, this isn't really blocking me, but it still seems worth taking a look at.

**Comment by icefoxen** [comment]
Can you tell whether the vector returned by `image.to_rgba8()` is empty?  Ie, is the problem the `to_rgba()` function, or the `Image::from_rgba8()` function?

**Comment by rocurley** [comment]
Ahh, should have thought of that:
```
pixels.iter().copied().all(|p| p == 0) = true
```
The issue is in `to_rgba`.

**Comment by IGBC** [comment]
confirmed on master branch with following snippet:

```use ggez::event::{self, EventHandler};
use ggez::{graphics, Context, ContextBuilder, GameResult};
use graphics::Drawable;

fn main() {
    // Make a Context.
    let (mut ctx, mut event_loop) = ContextBuilder::new("my_game", "Cool Game Author")
        .build()
        .expect("aieee, could not create ggez context!");

    // Create an instance of your event handler.
    // Usually, you should provide it with the Context object to
    // use when setting your game up.
    let mut my_game = MyGame::new(&mut ctx);

    // Run!
    event::run(ctx, event_loop, my_game);
}

struct MyGame {
    // Your state here...
}

impl MyGame {
    pub fn new(_ctx: &mut Context) -> MyGame {
        // Load/create resources such as images here.
        MyGame {
                    // ...
                }
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        // Update code here...
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, graphics::Color::WHITE);
        // Draw code here...
        let canvas = graphics::Canvas::with_window_size(ctx)?;
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_rectangle(
            ctx,
            graphics::DrawMode::fill(),
            graphics::Rect::new(10., 10., 100., 100.),
            graphics::Color::BLACK
        )?
        .draw(ctx, graphics::DrawParam::new())?;
        graphics::set_canvas(ctx, None);
        let image = canvas.into_inner();
        let image = break_everything(ctx, image)?; //Comment this line out to get expected behavior
        image.draw(ctx, graphics::DrawParam::new())?;
        graphics::present(ctx)
    }
}

fn break_everything(ctx: &mut Context, img: graphics::Image) -> GameResult<graphics::Image> {
    let w = img.width();
    let h = img.height();
    let pixels = img.to_rgba8(ctx)?;
    graphics::Image::from_rgba8(ctx, w, h, &pixels)
}
```

I will dig into to_rgba()

**Comment by PSteinhaus** [comment]
@IGBC Thanks! Good idea to look into this right now!
BUT I think you may want to look at the devel branch (generally, but especially here), since we just changed ```to_rgba()``` in #936 to no longer flip image data. Who knows, maybe we fixed this issue somehow as well (one is still allowed to dream, right? ^^)

**Comment by IGBC** [comment]
will do at next available opportunity

**Comment by PSteinhaus** [comment]
I just tried this in the current devel branch and I'm pretty confident that #936 fixed this problem.
Updated example code for the devel branch API:
```rust
use ggez::event::{self, EventHandler};
use ggez::{graphics, Context, ContextBuilder, GameResult};
use graphics::Drawable;
use ggez::graphics::Canvas;

fn main() -> GameResult {
    // Make a Context.
    let (mut ctx, mut event_loop) = ContextBuilder::new("my_game", "Cool Game Author")
        .build()
        .expect("aieee, could not create ggez context!");

    // Create an instance of your event handler.
    // Usually, you should provide it with the Context object to
    // use when setting your game up.
    let mut my_game = MyGame::new(&mut ctx);

    // Run!
    event::run(ctx, event_loop, my_game)
}

struct MyGame {
    my_canvas: Canvas,
}

impl MyGame {
    pub fn new(ctx: &mut Context) -> MyGame {
        // Create a canvas, draw something on it and then keep it
        let canvas = graphics::Canvas::with_window_size(ctx).unwrap();
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_rectangle(
            ctx,
            graphics::DrawMode::fill(),
            graphics::Rect::new(10., 10., 100., 100.),
            graphics::Color::BLACK,
        ).unwrap()
            .draw(ctx, graphics::DrawParam::new()).unwrap();
        graphics::present(ctx).unwrap(); // necessary to actually draw something on the canvas
        graphics::set_canvas(ctx, None);
        
        MyGame {
            my_canvas: canvas,
        }
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        // Update code here...
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        // Draw code here...
        graphics::clear(ctx, graphics::Color::WHITE);
        let image = self.my_canvas.to_image(ctx).unwrap();
        let image = break_everything(ctx, image)?; //Comment this line out to get expected behavior
        image.draw(ctx, graphics::DrawParam::new())?;
        graphics::present(ctx)
    }
}

fn break_everything(ctx: &mut Context, img: graphics::Image) -> GameResult<graphics::Image> {
    let w = img.width();
    let h = img.height();
    let pixels = img.to_rgba8(ctx)?;
    graphics::Image::from_rgba8(ctx, w, h, &pixels)
}
```
Whether I comment out the line containing ```break_everything``` or not I get the same behavior, there is even no flipping anymore, since #936 fixed that.

But please note that I moved the code drawing on the canvas to the constructor of MyGame and called ```graphics::present``` on it. This is necessary for flushing the draw buffer and actually drawing to it. You can also do all of that directly inside of the draw function, but then you'll get a flickering image, since you'll always have to call ```graphics::present``` twice in a row.

**Comment by IGBC** [comment]
Does the bug reappear if the canvas draw is made in the draw call? There is actually a valid reason do do this (mostly post processing effects).

Also prior to current dev drawing to a canvas did not require calling present, which again is required for smooth full screen post processing fx, we should probably check we haven't broken an API with current dev.

**Comment by IGBC** [comment]
Testing on `devel` with this code:
```rust
use ggez::event::{self, EventHandler};
use ggez::{graphics, Context, ContextBuilder, GameResult};
use graphics::Drawable;

fn main() {
    // Make a Context.
    let (mut ctx, mut event_loop) = ContextBuilder::new("my_game", "Cool Game Author")
        .build()
        .expect("aieee, could not create ggez context!");

    // Create an instance of your event handler.
    // Usually, you should provide it with the Context object to
    // use when setting your game up.
    let mut my_game = MyGame::new(&mut ctx);

    // Run!
    event::run(ctx, event_loop, my_game);
}

struct MyGame {
    // Your state here...
}

impl MyGame {
    pub fn new(_ctx: &mut Context) -> MyGame {
        // Load/create resources such as images here.
        MyGame {
		    // ...
		}
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        // Update code here...
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, graphics::Color::WHITE);
        // Draw code here...
        let canvas = graphics::Canvas::with_window_size(ctx)?;
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_rectangle(
            ctx,
            graphics::DrawMode::fill(),
            graphics::Rect::new(10., 10., 100., 100.),
            graphics::Color::BLACK
        )?
        .draw(ctx, graphics::DrawParam::new())?;
        graphics::present(ctx)?;
        graphics::set_canvas(ctx, None);
        let image = canvas.to_image(ctx)?;
        let image = break_everything(ctx, image)?; //Comment this line out to get expected behavior
        image.draw(ctx, graphics::DrawParam::new())?;
        graphics::present(ctx)
    }
}

fn break_everything(ctx: &mut Context, img: graphics::Image) -> GameResult<graphics::Image> {
    let w = img.width();
    let h = img.height();
    let pixels = img.to_rgba8(ctx)?;
    graphics::Image::from_rgba8(ctx, w, h, &pixels)
}
```

OK so with or without the present call before the render target swap, this has broken expected behavour. So I can conclude that #936 has broken real time re-rendering to canvases, which is a pretty big deal. You didn't implement it by rendering the canvas to the framebuffer did you?

**Comment by IGBC** [comment]
note to self: expected behavour returns in 743b40cdac94d79e419e289e534deccca9dcdf65 but patching `devel-head` back to lyon 16 does not fix it.

**Comment by PSteinhaus** [comment]
> So I can conclude that #936 has broken real time re-rendering to canvases, which is a pretty big deal.

I didn't even realize that real time re-rendering to canvases was a thing, but now that I know it used to work that way many example snippets suddenly make much more sense...

**Comment by nobbele** [comment]
uh what's real time re-rendering to canvas?

**Comment by IGBC** [comment]
its where you render part of your frame to a canvas. Move the canvas into texture memory then continue to render the frame to framebuffer. Mostly its used for full screen effects (bloom, motion blur, etc), or camera in camera effects (mirrors, or a simulated video screen that is looking somewhere else in the scene)

**Comment by nobbele** [comment]
Hm I see yeah

**Comment by IGBC** [comment]
GGEZ being a 2D engine, its important we keep support for this technique as its one of the few ways a lot of artistic effects in 2D games can be implemented. 

Also Modern GPU's are measured in TFLOPs now and as A 2D engine we gotta use them somehow :P

**Comment by nobbele** [comment]
yes but how, I don't even see how the PR would change the behavior of that but as always with canvas who knows what's happening.

**Comment by IGBC** [comment]
> yes but how, I don't even see how the PR would change the behavior of that but as always with canvas who knows what's happening.

yea in my testing this morning I determined I was mistaken, its not #936, it was broken before that. The last commit in the devel that I have tested it with and it works is 743b40cdac94d79e419e289e534deccca9dcdf65 so the regression is somewhere between there and devel-head

**Comment by nobbele** [comment]
ah right, can you try to determine the exact one? (try something inbetween like ee00345b19e2b538db0e0516d5b33ced32331600)
I tried the code you sent but both devel and 743b40c just blink a lot so I don't know lol

**Comment by IGBC** [comment]
> I tried the code you sent but both devel and [743b40c](https://github.com/ggez/ggez/commit/743b40cdac94d79e419e289e534deccca9dcdf65) just blink a lot

did you remember to remove the present call in the canvas drawing?

**Comment by IGBC** [comment]
> ah right, can you try to determine the exact one? (try something inbetween like [ee00345](https://github.com/ggez/ggez/commit/ee00345b19e2b538db0e0516d5b33ced32331600))

I can but I am going to have to slowly test it with every commit, so it'll take me a while. I will check back in on Saturday maybe, Last exam is Friday, so I don't see me having time to thoroughly hunt this down until Saturday. 

I am gonna notify @icefoxen about this bug, and promote it to a 0.6 blocker, because from my viewpoint this is quite serious.

**Comment by nobbele** [comment]
No need to test every commit, just do one half of the range at a time. Do the exams first though yeah

**Comment by IGBC** [comment]
Binary search? By hand... Damn, reminding me of different exams from long ago XD

**Comment by nobbele** [comment]
Bad commit is 4f27eb5b0b7549db270865ea3d011ac5712af339, as I expected

**Comment by nobbele** [comment]
`Canvas.to_rgba8()` has never worked without `graphics::present`, only difference now is that we use it for `Canvas.to_image()`.
Do we investigate and fix to_rgba8 or add a note to `Canvas.to_image()` that real time canvas rendering requires managing `Canvas.raw_image()` manually?

**Comment by nobbele** [comment]
I am so done with canvases. After testing I have gotten these two test codes
```rust
graphics::clear(ctx, graphics::Color::BLACK);
        let canvas = graphics::Canvas::with_window_size(ctx)?;
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_circle(
            ctx,
            DrawMode::fill(),
            vec2(0.0, 0.0),
            10.0,
            1.0,
            Color::WHITE,
        )?
        .draw(ctx, DrawParam::new())?;
        graphics::present(ctx)?;
        graphics::set_canvas(ctx, None);
        canvas.draw(ctx, DrawParam::new())?;
        canvas
            .raw_image()
            .draw(ctx, DrawParam::new().dest(vec2(0.0, 0.0)))?;
        canvas
            .to_image(ctx)?
            .draw(ctx, DrawParam::new().dest(vec2(15.0, 15.0)))?;
        graphics::present(ctx)?;
```
above one blinks (double present) but renders properly
```rust
graphics::clear(ctx, graphics::Color::BLACK);
        let canvas = graphics::Canvas::with_window_size(ctx)?;
        graphics::set_canvas(ctx, Some(&canvas));
        graphics::Mesh::new_circle(
            ctx,
            DrawMode::fill(),
            vec2(0.0, 0.0),
            10.0,
            1.0,
            Color::WHITE,
        )?
        .draw(ctx, DrawParam::new())?;
        graphics::set_canvas(ctx, None);
        canvas.draw(ctx, DrawParam::new())?;
        canvas
            .raw_image()
            .draw(ctx, DrawParam::new().dest(vec2(0.0, 0.0)))?;
        canvas
            .to_image(ctx)?
            .draw(ctx, DrawParam::new().dest(vec2(15.0, 15.0)))?;
        graphics::present(ctx)?;
```
above one renders all but the canvas.to_image() one. Only change is graphics::present() at the end of the canvas render

**Comment by PSteinhaus** [comment]
> `Canvas.to_rgba8()` has never worked without `graphics::present`, only difference now is that we use it for `Canvas.to_image()`.
> 
> Do we investigate and fix to_rgba8 or do we add a note to `Canvas.to_image()` that real time canvas rendering requires managing `Canvas.raw_image()` manually?

Definitely not the latter. This stuff used to work, so I'm all in for fixing it again. Why doesn't `Canvas.to_rgba8()` work without `graphics::present`?

**Comment by nobbele** [comment]
fyi the reason that commit is the bad one is because that's when to_image was switched to using to_rgba8. to_rgba8 never worked

**Comment by PSteinhaus** [comment]
> above one renders all but the canvas.to_image() one. Only change is graphics::present() at the end of the canvas render

Now I'm completely confused.

The broken behavoir _should_ be that NOTHING is rendered, because real-time-rendering is broken and the canvas has not yet been drawn to... 

Also, @IGBC already checked and realized that #936 [isn't at fault for this regression](https://github.com/ggez/ggez/issues/813#issuecomment-853368957).

So what you've found here sounds more like a different bug that might be connected...

**Comment by PSteinhaus** [comment]
Ok, so I checked some pre#936 builds and: they actually DO NOT show this kind of regression!
So I guess it is #936's fault after all.

Moreover, it turns out **the regression wasn't real at all**. The thing that introduces the bug is `Canvas::to_image`, as nobbele correctly pointed out. I checked, and if one uses `Canvas::draw` or `draw` on `Canvas::raw_image` then real time painting on canvas works as it should, again just like nobbele already said.

So it seems the problem lays in that the serializing of the image happens _before_ anything is drawn to it in the current frame.
